File: daterem.pl
#!/usr/bin/perl -w
use strict;
use Time::Local;

my ($rday,$rmonth,$ryear);
my $day = 60*60*24;
my ($unter,$ober);
my $easter;
my @alles;

###########
sub options
###########
{
	my ($day,$month,$year);
	if (@ARGV>=2) {
		print "\nUsage: $0 [[[dd.]mm.]yyyy]\n\n";
		exit 1;
	} 
	if (@ARGV==1) {
		($day,$month,$year) = split(/\./,$ARGV[0]);
		if (! defined $year) {
			$year = $month;
			$month = $day;
			undef $day;
		}
		if (! defined $year) {
			$year = $month;
			undef $month;
		}
	} else {
		($day,$month,$year) = (localtime)[3,4,5];
		$month++;
		$year += 1900;
	}
	return ($day,$month,$year);
}
##############
sub calceaster # Ostern berechnen
##############
{
	my $j = $ryear;
	my $a = $j % 19;
	my $b = int($j / 100);
	my $c = $j % 100;
	my $d = int($b / 4);
	my $e = $b % 4;
	my $f = int(( $b + 8 ) / 25);
	my $g = int(( $b - $f + 1 ) / 3);
	my $h = ( 19 * $a + $b - $d - $g + 15 ) % 30;
	my $i = int($c / 4);
	my $k = $c % 4;
	my $l = ( 32 + 2 * $e + 2 * $i - $h - $k ) % 7;
	my $m = int(( $a + 11 * $h + 22 * $l ) / 451);
	my $n = int(( $h + $l - 7 * $m + 114 ) / 31);
	my $o = ( $h + $l - 7 * $m + 114 ) % 31;

	return timelocal(0,0,12,$o+1,$n-1,$j-1900);
}
#############
sub vergleich # Vergleichsfunktion um zwei Datumszeilen zu vergleichen
#############
{
	my (@ha,@hb);
	@ha = split(/ /,$a);
	@ha = split(/\./,$ha[0]);
	@hb = split(/ /,$b);
	@hb = split(/\./,$hb[0]);
	return timelocal(0,0,12,$ha[0],$ha[1]-1,$ha[2]-1900) 
               <=>
	       timelocal(0,0,12,$hb[0],$hb[1]-1,$hb[2]-1900);
}
###############
sub zeitstempel
###############
{
	my $zeile = $_[0];
	$zeile =~ s/([0-9.]+)\s.*/$1/;
	my ($day,$month,$year) = split(/\./,$zeile);
	return timelocal(0,0,12,$day,$month-1,$year-1900);
}
#############
sub wochentag
#############
{
	my @wochentag = qw(So Mo Di Mi Do Fr Sa);
	return $wochentag[
		(localtime(
			zeitstempel($_[0])
		))[6]
	];
}
############
sub add2list
############
{
	my ($dd,$mm,$yyyy,$remark) = @_;
	$dd = "0" . $dd if (($dd < 10) && (length($dd)<2));
	$mm = "0" . $mm if (($mm < 10) && (length($mm)<2));
	push @alles,"$dd.$mm.$yyyy $remark";
}
###########
sub readdat
###########
{
	my $zeit;
	my $beschreibung;
	my $dat1;
	my $dat2;
	my ($dat1_day,$dat1_month,$dat1_year);
	my ($dat2_day,$dat2_month,$dat2_year);
	if (-r 'daterem.dat') {
		open(DATA,"< daterem.dat") 
			or die "\nIrgendetwas stimmt hier nicht!\n\n";
		while (<DATA>) {
			chomp;
			s/^#.*//g; # entfernt Kommentarzeilen
			s/\s{2,}/ /g; # ersetzt mehrere WS durch ein Leerzeichen
			s/^\s+//g; # entfernt Whitespace(s) am Zeilenanfang
			s/\s+$//g; # entfernt Whitespace(s) am Zeilenende
			if (m/^.+$/) {
				($zeit,$beschreibung) = split(/ /, $_, 2);
				if ($zeit =~ m/^[-0-9]+$/) { # Osterdaten
					$zeit = $zeit * $day + $easter;
					($dat1_day,$dat1_month,$dat1_year) = (localtime($zeit))[3,4,5];
					$dat1_month++;
					$dat1_year += 1900;
					add2list($dat1_day,$dat1_month,$dat1_year,$beschreibung);
				} elsif ($zeit =~ m/.+-.+/) { # Ein Zeitraum
					($dat1,$dat2) = split(/-/ , $zeit);
					($dat1_day,$dat1_month,$dat1_year) = split(/\./,$dat1);
					($dat2_day,$dat2_month,$dat2_year) = split(/\./,$dat2);
					$dat2_year ||= $ryear;
					$dat1_month ||= $dat2_month;
					$dat1_year ||= $dat2_year;
					
					$dat1 = timelocal(0,0,12,$dat1_day,$dat1_month-1,$dat1_year-1900);
					$dat2 = timelocal(0,0,12,$dat2_day,$dat2_month-1,$dat2_year-1900);
					while ($dat1 <= $dat2) {
						($dat1_day,$dat1_month,$dat1_year) = (localtime($dat1))[3,4,5];
						$dat1_month++;
						$dat1_year += 1900;
						add2list($dat1_day,$dat1_month,$dat1_year,$beschreibung);
						$dat1 += $day;
					}
					
				} else { # Ein einzelnes Datum
					($dat1_day,$dat1_month,$dat1_year) = split(/\./,$zeit);
					$dat1_year ||= $ryear;
					add2list($dat1_day,$dat1_month,$dat1_year,$beschreibung);
				}
			}
		}
		close DATA;
	} else {
		print "\nNo datafile!\n\n";
	}
}

($rday,$rmonth,$ryear) = options;
$easter = calceaster;
readdat;
@alles = sort vergleich @alles;

if (! defined $rmonth) {
	foreach (grep(/\.$ryear /,@alles)) {
		print wochentag($_),", $_\n";
	}
} elsif (! defined $rday) {
	foreach (grep(/\.$rmonth\.$ryear /,@alles)) {
		print wochentag($_),", $_\n";
	}
} else {
	$unter = timelocal(0,0,0,$rday,$rmonth-1,$ryear-1900);
	$ober  = timelocal(59,59,23,$rday,$rmonth-1,$ryear-1900);
	foreach (@alles) {
		if ((zeitstempel($_)>=$unter) && (zeitstempel($_)<=$ober)) {
			print wochentag($_),", $_\n";
		}
	}
}

File: daterem.py
#!/usr/bin/env python3

import sys
import re
import time
import argparse


def usage():
    print("\nUsage: %s [[[dd.]mm.]yyyy]\n\n" % sys.argv[0])


def options():
    global filename
    global born_list
    global dead_list

    parser = argparse.ArgumentParser(
        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
        description="A simple date reminder.",
    )

    parser.add_argument("-f", "--file", help="file to read from", default="daterem.dat")
    parser.add_argument(
        "-b", "--born", help="string list indicating a born year", default="born,year"
    )
    parser.add_argument(
        "-d",
        "--dead",
        help="string list indicating a start year",
        default="dead,started",
    )
    parser.add_argument(
        "date",
        help="search for a specific date [[[dd.]mm.]yyyy]",
        nargs="?",
        default="today",
    )
    args = parser.parse_args()

    filename = args.file
    born_list = args.born.split(",")
    dead_list = args.dead.split(",")

    if args.date == "today":
        oday, omonth, oyear = map(time.strftime, ["%d", "%m", "%Y"])

    else:
        optionlist = args.date.split(".")

        if len(optionlist) == 1:
            oyear = optionlist[0]
            omonth = ""
            oday = ""
        elif len(optionlist) == 2:
            oyear = optionlist[1]
            omonth = optionlist[0]
            oday = ""
        else:
            oyear = optionlist[2]
            omonth = optionlist[1]
            oday = optionlist[0]

        if len(oday) == 1:
            oday = "0" + oday

        if len(omonth) == 1:
            omonth = "0" + omonth

    return (oday, omonth, oyear)


def to_epoch(date):
    date = date.split()
    return time.mktime(time.strptime(date[0] + " 12", "%d.%m.%Y %H"))


def to_date(epoch):
    return time.strftime("%d.%m.%Y", time.localtime(epoch))


def calceaster():  # Calculate easter date
    # https://en.wikipedia.org/wiki/Computus#Anonymous_Gregorian_algorithm
    j = int(ryear)
    a = j % 19
    b = int(j / 100)
    c = j % 100
    d = int(b / 4)
    e = b % 4
    f = int((b + 8) / 25)
    g = int((b - f + 1) / 3)
    h = (19 * a + b - d - g + 15) % 30
    i = int(c / 4)
    k = c % 4
    l = (32 + 2 * e + 2 * i - h - k) % 7
    m = int((a + 11 * h + 22 * l) / 451)
    n = int((h + l - 7 * m + 114) / 31)
    o = (h + l - 7 * m + 114) % 31 + 1

    return to_epoch("%s.%s.%s" % (str(o), str(n), str(j)))


def readdat():
    try:
        f = open(filename, "r")
    except:
        print("Error: cannot open file '%s'" % filename)
        sys.exit(1)

    for line in f:
        line = line.strip()

        if not re.match("^#", line):
            line = re.sub(r"\s{2,}", " ", line)
            ltime, description = line.split(" ", 1)

            if re.match("^[-0-9]+$", ltime):  # Datum abhaengig von Ostern
                ltime = int(ltime) * day + easter
                alldates.append("%s %s" % (to_date(ltime), description))
            elif re.match(".+-.+", ltime):  # Zeitraeume
                description = description + " [" + ltime + "]"
                dat1, dat2 = ltime.split("-")
                dat1list = dat1.split(".")
                dat2list = dat2.split(".")

                if len(dat1list) == 2:
                    dat1list.append("")

                if len(dat2list) == 2:
                    dat2list.append("")

                if not dat2list[2]:
                    dat2list[2] = ryear

                if not dat1list[2]:
                    dat1list[2] = dat2list[2]

                    if not dat1list[1]:
                        dat1list[1] = dat2list[1]

                dat1 = to_epoch(dat1list[0] + "." + dat1list[1] + "." + dat1list[2])
                dat2 = to_epoch(dat2list[0] + "." + dat2list[1] + "." + dat2list[2])

                while dat1 <= dat2:
                    alldates.append("%s %s" % (to_date(dat1), description))
                    dat1 = dat1 + day

            else:  # Ein einzelnes Datum
                ltimelist = ltime.split(".")
                ltimelist.append("")

                if ltimelist[2] == "":
                    ltimelist[2] = str(int(ryear) + 1)
                    if "countdown" in description:
                        alldates.append(
                            "%s.%s.%s %s"
                            % (ltimelist[0], ltimelist[1], ltimelist[2], description)
                        )
                    ltimelist[2] = ryear

                alldates.append(
                    "%s.%s.%s %s"
                    % (ltimelist[0], ltimelist[1], ltimelist[2], description)
                )

    f.close()


def printline(line):
    date, text = line.split(" ", 1)
    prt = time.strftime("%a, %d.%m.%Y", time.localtime(to_epoch(date))) + ", " + text

    # Default is born/year for born_list and dead/started for dead_list
    matching_list = born_list + dead_list
    matching_strings = str.join("|", matching_list)
    match = re.match(r".*(" + matching_strings + r") (\d{4})", text)

    if match:

        if match.group(1) in (born_list):
            prt += ", age "
        elif match.group(1) in (dead_list):
            prt += ", ago: "

        age = int(ryear) - int(match.group(2))
        prt += str(age) + " year"

        if age > 1:
            prt += "s"

    if ", countdown" in line:
        duration = int(
            (to_epoch(date) - to_epoch(time.strftime("%d.%m.%Y"))) / day + 0.5
        )

        newmatch = re.search(r"countdown-([0-9]+)", line)
        if newmatch:
            daysbefore = int(newmatch.group(1))
            line = line.replace("countdown-90", "countdown")
        else:
            daysbefore = 365

        if duration == 0:
            countdown = " today"
        elif duration == 1:
            countdown = " tomorrow"
        elif 1 < duration <= daysbefore:
            countdown = " in %s days" % duration
        else:
            countdown = ""
            prt = ""

        prt = prt.replace(" countdown-" + str(daysbefore), countdown)
        prt = prt.replace(" countdown,", countdown)
        prt = prt.replace(" countdown", countdown)

    if prt:
        print(prt)


def main():
    global rday  # Reference Day (if any)
    global rmonth  # Reference Month (if any)
    global ryear  # Reference Year
    global day  # One day in seconds
    global easter  # Easter date - 12:00 - in seconds since the Epoch
    global alldates  # List containing all dates
    global filename  # The datafile (usually "daterem.dat")

    alldates = []
    day = 60 * 60 * 24

    rday, rmonth, ryear = options()
    easter = calceaster()
    readdat()
    alldates = sorted(alldates, key=to_epoch)

    if not rmonth:
        search = ".%s " % ryear
    elif not rday:
        search = ".%s.%s " % (rmonth, ryear)
    else:
        search = "%s.%s.%s " % (rday, rmonth, ryear)

    for line in alldates:
        if search in line:
            printline(line)
        elif ", countdown" in line:
            printline(line)


if __name__ == "__main__":
    main()
